src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1529,6 +1529,89 @@ JvmtiEnvBase::is_in_thread_list(jint count, const jthread* list, oop jt_oop) {
   return false;
 }
 
+class VM_SetNotifyJvmtiEventsMode : public VM_Operation {
+private:
+  bool _enable;
+
+  // Iterates over all JavaThread's, counts VTMS transitions and restores
+  // jt->jvmti_thread_state() and jt->jvmti_vthread() for VTMS transition protocol.
+  int count_transitions_and_correct_jvmti_thread_states() {
+    int count = 0;
+
+    for (JavaThread* jt : ThreadsListHandle()) {
+      oop  ct_oop = jt->threadObj();
+      oop  vt_oop = jt->vthread();
+      JvmtiThreadState* jt_state = jt->jvmti_thread_state();
+      JvmtiThreadState* ct_state = java_lang_Thread::jvmti_thread_state(jt->threadObj());
+      JvmtiThreadState* vt_state = vt_oop != nullptr ? java_lang_Thread::jvmti_thread_state(vt_oop) : nullptr;
+      bool virt = vt_oop != nullptr && vt_oop != ct_oop;
+
+      if (jt->is_in_VTMS_transition()) {
+        count++;
```

Comment:
Nice catch, thanks! Fixed now.

Reply:
The issue I see is that during a transition the vt_oop that we'll get will almost always be that of the carrier thread, because it's almost the first thing we change when unmounting, and the last we change when mounting.